### PR TITLE
Squash all apt installation commands into a single command

### PR DESCRIPTION
**Pull request by emersonknapp**
Fixes https://github.com/ros-tooling/cross_compile/issues/165

See the issue for the rationale behind making this change

Signed-off-by: Emerson Knapp <emerson.b.knapp@gmail.com>

### CODE REVIEW

File: ros_cross_compile/docker/gather_rosdeps.sh [hunk 1]
```diff
@@ -34,7 +34,18 @@ rosdep install \
     --default-yes \
     --skip-keys "${SKIP_ROSDEP_KEYS}" \
     --simulate \
-  >> "${OUT_PATH}"
+  >> /tmp/all-deps.sh
+
+# Find the non-apt lines and move them as-is to the final script
+grep -v "apt-get install -y" /tmp/all-deps.sh >> "${OUT_PATH}" || true
```
Comment (thomas-moulard) [inline ros_cross_compile/docker/gather_rosdeps.sh:40]: Why `|| true` ?

File: ros_cross_compile/docker/gather_rosdeps.sh [hunk 2]
```diff
@@ -34,7 +34,18 @@ rosdep install \
...
+
+# Find all apt-get lines from the rosdep output
+# As an optimization, we will combine all such commands into a single command to save time
+grep "apt-get install -y" /tmp/all-deps.sh > /tmp/apt-deps.sh || true
```
Comment (thomas-moulard) [inline ros_cross_compile/docker/gather_rosdeps.sh:44]: ditto

**Comment by thomas-moulard** [comment]
How much time are we saving here? This code will break if rosdep is updated i.e. `-y` => `--yes`. There is a trade-off between maintenance burden and speed here, it'd be good to make a data based decision.

Also, if we could get `rosdep` to output a list of packages to avoid parsing the command that'd be perfect. Do you know if it's possible?

**Comment by emersonknapp** [comment]
I can find data later - anecdotally it saves me several minutes, i think more than 10, when this step has to be re-run on my workspace.

The output format is unlikely to change anytime soon, it's been stable for years. However, if the format does change such that the grep commands don't match anymore, the functionality will continue working, just this optimization will not happen.

The rosdep CLI doesn't have that kind of output and I'm hesitant to use the Python API https://docs.ros.org/independent/api/rosdep/html/rosdep2_api.html because it has limited documentation and is explicitly marked unstable. A python script based on the rosdep api, instead of this docker container, may be a long-term desirable outcome, though.

**Comment by thomas-moulard** [comment]
Thanks for explaining. It's interesting to see that it is that much faster.

**Pull request merged**